sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java

```diff
@@ -189,7 +189,14 @@ public int read(ByteBuffer dst) throws IOException {
       if (closed) {
         throw new IOException("Channel is closed");
       }
-      return inputStream.read(dst);
+      // We avoid using the ByteBuffer based read for Hadoop because some FSInputStream
+      // implementations are not ByteBufferReadable,
+      // See https://issues.apache.org/jira/browse/HADOOP-14603
+      int read = inputStream.read(dst.array());
```

Comment:
@lukecwik Not sure if I understand, if you look at TextSource, it defines a fixed size for the read buffer, so we have a limit. TextBasedReader already defines it to be 8192.
https://github.com/apache/beam/blob/5181e619f17e1f69fabe8d5bdfc7a3a6a2142cde/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java#L89-L90

And the method that iterates to read (by delegating to the HadoopSeekableByteChannel) is already moving the buffer start position to zero on each iteration so we cannot get off heap, no?
https://github.com/apache/beam/blob/5181e619f17e1f69fabe8d5bdfc7a3a6a2142cde/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java#L231